span.go

```diff
@@ -6,10 +6,52 @@ import (
 	"time"
 )
 
+type SpanContext interface {
+	// SetBaggageItem sets a key:value pair on this Span that also
+	// propagates to future Span children.
+	//
+	// SetBaggageItem() enables powerful functionality given a full-stack
+	// opentracing integration (e.g., arbitrary application data from a mobile
+	// app can make it, transparently, all the way into the depths of a storage
+	// system), and with it some powerful costs: use this feature with care.
+	//
+	// IMPORTANT NOTE #1: SetBaggageItem() will only propagate trace
+	// baggage items to *future* children of the Span.
+	//
+	// IMPORTANT NOTE #2: Use this thoughtfully and with care. Every key and
+	// value is copied into every local *and remote* child of this Span, and
+	// that can add up to a lot of network and cpu overhead.
+	//
+	// IMPORTANT NOTE #3: Baggage item keys have a restricted format:
+	// implementations may wish to use them as HTTP header keys (or key
+	// suffixes), and of course HTTP headers are case insensitive.
+	//
+	// As such, `restrictedKey` MUST match the regular expression
+	// `(?i:[a-z0-9][-a-z0-9]*)` and is case-insensitive. That is, it must
+	// start with a letter or number, and the remaining characters must be
+	// letters, numbers, or hyphens. See CanonicalizeBaggageKey(). If
+	// `restrictedKey` does not meet these criteria, SetBaggageItem()
+	// results in undefined behavior.
+	//
+	// Returns a reference to this Span for chaining, etc.
+	SetBaggageItem(restrictedKey, value string) SpanContext
```

Comment:
I really like how the baggage stuff moves to `SpanContext`... much clearer, conceptually speaking.

I didn't change the comments at all above... obv it looks like we'll move to remove these restrictions, but that's orthogonal.